### ISSUE TITLE
Interfaces for style

### ISSUE THREAD

**Issue by ogoffart** [slint-ui/slint#1870]
Currently, the style must re-implement a lot of components, but there is nothing that checks that these component have the same interface/properties.

The idea would be to build interfaces like so:
(based on the current interface of Button https://docs.rs/slint/latest/slint/docs/widgets/index.html#button)

```slint
interface Button {
   in property <string> text;
   in property <bool> checkable;
   in-out property <string> checked;
   out property <bool> pressed;
   in property <image> icon;
   callback clicked();
}
```

Then, the style would have to implement this interface:

```
// FIXME should we import that? from where, should it be bluit-in
import { Button } from "std.slint"; 

component MyStyleButton implements Button {
   pressed: ta.pressed; // can set output properties
   Text { text: root.text } // can access property
   ta := TouchArea { clicked => { root.clicked() } } // call callbacks
   
   in property <int> extra_prop; // ERROR!  not allowed to add public properties
}

// re-export the MyStyleButton as the Button of our style
export MyStyleButton as Button;
```

It is a bit akward to have to go through `MyStyleButton` as a separate name, maybe we could allow `component Button implements Button {...}`  and have different namespace for components and for interface.

Unresolved question:
 - How do we make sure that the style implements all the widgets?
 - How do we make sure that a style don't export more than the required widget?
 - Is it a breaking change, in Slint, to add more widgets?  Should we add some sort of versioning?
 - Should the interface allow to also have some logic in them?
 - What about widgets like TabWidget or ListView that needs compiler support?
 - Should we even allow users to declare new interface, or are these just be builtin?

**Comment by tronical** [comment]
> // FIXME should we import that? from where, should it be bluit-in
> import { Button } from "std.slint"; 

Perhaps a dedicated file makes sense. For example `std-widget-interfaces.slint`

Regarding the names, we could also do something like this:

```
import * as Interfaces from "std-widget-interfaces.slint";

component Button implements Interfaces.Button {
}
```

**Comment by tronical** [comment]
> How do we make sure that a style don't export more than the required widget?

We could say that once a component implements an interface, only that interface is visible. Only components that don't implement an interface implicitly define an interface (with the same name).

**Comment by Enyium** [comment]
`interface` reminds me of Java's interfaces and Rust's traits, where you can implement *multiple* for a given blueprint (`class`/`struct`), and where you need to redundantly repeat the interface content again to make it concrete. I'm not so sure this would be fitting. Alternative terms: `connector`, `bond`.

I think what you currectly call `interface` would be included in the public API of the component, right? Then, even if you wanted to implement a widget in multiple subparts and then present it as a whole, it would make sense to only allow one `interface`. But to accommodate this kind of region-wise internal implementation, it could be useful, if there was something like the possibility to spread out (or flatten) an interface *in* another interface:

```slint
export interface WidgetInterface {
    spread interface <AInterface> a-interface;
    spread interface <BInterface> b-interface;
}

interface AInterface {
    in-out property <int> foo;
}

interface BInterface {
    in-out property <string> bar;
}

export component Widget implements WidgetInterface {
    a-interface <=> a.interface;
    b-interface <=> b.interface;

    a := A {}
    b := B {}
}

component A implements AInterface {
    foo: 123;
}

component B implements BInterface {
    bar: "abc";
}
```

**Comment by jwtocadero** [comment]
@tronical I think that not allowing extra public properties is a mistake. Sometimes we don't want to add just style, but also behavior. Like an on off button which has a `cancelling` state. this state might be set from the outside world. I think interfaces should be implemented like Go does it. You define a set of callbacks, properties etc and whatever object has them does is qualified as implementation of the interface. No need to limit the implementation to do more in a context which is aware of the actual type and not just the interface.

**Comment by 0x6e** [comment]
>> How do we make sure that a style don't export more than the required widget?
>
> We could say that once a component implements an interface, only that interface is visible. Only components that don't implement an interface implicitly define an interface (with the same name).

I don't think it is useful to enforce that only the interface API is exposed. I can see why it might be desirable from a std-widgets point of view: we don't want someone implementing a std-widgets style to add extra API that isn't compatible with other existing styles.

On the other hand I might want to use e.g. the LineEdit interface to produce my own SearchLineEdit. This wouldn't be part of a std-widgets style, but I want to re-use the line edit interface and then maybe also add my own additional API.

Perhaps for the std-widgets use case what we actually need is to enforce/warn about additional API at the compiler/internal library level.

In the case where I might be able to use the interface as a property type I wouldn't expect to be able to access the non-interface API anyway:
```
component Foo {
    in property<MyInterface> my-interface;
}
```

**Comment by tronical** [comment]
Further discussion between @ogoffart, me, and @0x6e brought up the following idea to permit composition for sharing functionality, but re-use the interface concept to conveniently define the API:

```slint
interface LineEditInterface {
    in-out property <string> text;
    ...
}

component LineEditBase implements LineEditInterface {
    // business logic here
}

// The final widget

export component LineEdit
    uses { LineEditInterface from base }
    inherits Something // optional
{
    base := LineEditBase { ... }

    accessibility-delegate: base;
    forward-focus: base;
}

```

**Comment by 0x6e** [comment]
I assume that we also want to be able to do:

```slint
component LineEditBase
    implements LineEditInterface
    inherits Something // optional
{
    // business logic here
}
```

**Comment by ogoffart** [comment]
I'm not sure we really need both `inherits` and (`implements` or `use`) at the same time.
I.e: i'd leave the `inherits` out at first and we can still consider it later.

**Comment by 0x6e** [comment]
I think the initial support for interfaces is complete enough to look at how this applies to std-widgets. I've submitted a draft PR/RFC looking at the LineEdit interface here: https://github.com/slint-ui/slint/pull/10722

**Comment by 0x6e** [comment]
Non-exhaustive list of tasks that I think are outstanding for interfaces as a non-experimental feature:

- [ ] Support for interface properties/declarations/functions to be provided by a component's derived type. Currently the interface must be implemented by the component with the `implements` statement;
- [ ] Verification that the API does what we need for std-widgets;
- [ ] Documentation;

**Comment by 0x6e** [comment]
One question that has come up whilst implementing the Button interface is how accessibility should be handled. For example, in `internal/compiler/widgets/material/button.slint` it would make sense if `MaterialButtonBase` implemented `Button` - it already has most of the properties. However, the tests fail because `Button` component does:

```slint
    accessible-action-default => {
        base.clicked();
    }
```

which bypasses the checked logic of `MaterialButtonBase`.

Full code and test output can be viewed [here](https://github.com/slint-ui/slint/commit/c7aaf8de6ac750e46ccebe8518bc9e22bc15cf00).

One solution was to duplicate the accessibility properties on the `MaterialButtonBase` and forward everything from the `Button` component through to the `base` element. However, we can't forward everything:

> The `accessible-role` property must be a constant expression

[This works](https://github.com/slint-ui/slint/commit/b682b7025a01bb21936c69162f4de6fac216d8f8), but it raises some questions:

1. Should the accessibility properties be part of the widget interfaces? (I think probably not?)
2. Should the accessibility properties be an interface in their own right?
 
Then we could have `MaterialButtonBase` implement that as well (would need to add support for implementing more than one interface) and then button could `uses { Accessibility from base, ButtonInterface from base }`.

We might have to relax the `accessible-role` property must be a constant expression rule, or do some custom interface-uses implementation for accessibility that copies the property declaration from the implementation to the "user", instead of creating a binding.

**Comment by tronical** [comment]
My guts feeling is that the accessibility properties are not part of the interface. They are an interface themselves - kind of.

> which bypasses the checked logic of MaterialButtonBase

To me it looks like `callback clicked <=> state-layer.clicked;` is missing from `MaterialButtonBase`. AFAICS the callback is in the interface, but it's not handled in `MaterialButtonBase`. I'm probably missing something though? :)

**Comment by 0x6e** [comment]
> To me it looks like callback clicked <=> state-layer.clicked; is missing from MaterialButtonBase. AFAICS the callback is in the interface, but it's not handled in MaterialButtonBase. I'm probably missing something though? :)

Ah yes, but you can't do that because it introduces a duplicate callback declaration:

```slint
    // error: Duplicated callback declaration
    // callback clicked <=> state-layer.clicked;
    //          ^^^^^^^
```

The callback declaration is copied from the interface into the implementing component.

Maybe that can be fixed in interfaces by applying interface callbacks _after_ parse the document, check the parameter and return value types?

However, I find it slightly odd that you can't do this, which you can do for properties:

```slint
    // error: 'clicked' is a callback. Use `=>` to connect
    // clicked <=> state-layer.clicked;
    // ^^^^^^^
```

I wasn't sure if there was an existing reason that you can't do that?

This is syntactically correct `clicked => state-layer.clicked();` but not what we want: `state-layer.clicked` is now the handler of our interface callback.

The solution was to invoke `root.clicked` from the `StateLayer` `clicked` callback.

**Comment by tronical** [comment]
> > To me it looks like callback clicked <=> state-layer.clicked; is missing from MaterialButtonBase. AFAICS the callback is in the interface, but it's not handled in MaterialButtonBase. I'm probably missing something though? :)
> 
> Ah yes, but you can't do that because it introduces a duplicate callback declaration:
> 
>     // error: Duplicated callback declaration
>     // callback clicked <=> state-layer.clicked;
>     //          ^^^^^^^
> The callback declaration is copied from the interface into the implementing component.
> 
> Maybe that can be fixed in interfaces by applying interface callbacks _after_ parse the document, check the parameter and return value types?

Right, re-declaring is not good. I think it's correct that this is an error.

> However, I find it slightly odd that you can't do this, which you can do for properties:
> 
>     // error: 'clicked' is a callback. Use `=>` to connect
>     // clicked <=> state-layer.clicked;
>     // ^^^^^^^
> I wasn't sure if there was an existing reason that you can't do that?

Ah yes, that's what I had in mind. Brainfart, that doesn't work indeed :)
 
> This is syntactically correct `clicked => state-layer.clicked();` but not what we want: `state-layer.clicked` is now the handler of our interface callback.
> 
> The solution was to invoke `root.clicked` from the `StateLayer` `clicked` callback.

I'm slightly confused now :). If we want to permit the accessibility handler in `Button` to invoke `base.clicked()` and `MaterialButtonBase` has `clicked => { state-layer.clicked(); }`, wouldn't the call chain work as expected?

**Comment by 0x6e** [comment]
Sorry, for the confusion. The current call-chain does work. What I don't like is that we now have to manually forward the accessibility API to base. I think the solution is to have the accessibility API as an interface, but one that requires special handling:

> We might have to relax the "accessible-role property must be a constant expression rule", or do some custom interface-uses implementation for accessibility that copies the property declaration from the implementation to the "user", instead of creating a binding.

**Comment by tronical** [comment]
Ahh, thanks for clearing that up :).

I think we've had discussions before about a generic `accessibility-delegate` type of property/concept, but my feeling is that this is unrelated to interfaces.

**Comment by 0x6e** [comment]
Ok, thank you. For now I'll leave the accessibility API alone.

**Comment by 0x6e** [comment]
There are a few cases where I'm not sure what the desired approach should be.

1. ListView inherits ScrollView, StandardListView & StandardTableView inherit ListView and add their own properties/callbacks/etc;
2. Grid-, Horizontal- and VerticalBox all inherit from their non-std-widgets base components and apply style-specific defaults.

For (1) I think we need to have a discussion about interface inheritance. So far an interface cannot `inherit` or `implements` another interface. Component inheritance still works, of course. ScrollView can implement ScrollViewInterface and that would be accessible to ListView. I think we have two options:

a) allow interface inheritance:

```slint
export interface ScrollView {
  // ScrollView API
}

export interface ListView inherits ScrollView {
  // Empty, but in the future we can use this to require that a std-widgets style implementation must provide a ListView component
}

export interface StandardListView inherits ListView {
  // Additional StandardListView API
}

export interface StandardTableView inherits ListView {
  // Additional StandardTableView API
}
```

b) require explicit interfaces:

```slint
export interface ScrollView {
  // ScrollView API
}

export interface ListView {
  // All ScrollView API
}

export interface StandardListView {
  // All ScrollView API
  // Additional StandardListView API
}

export interface StandardTableView {
  // All ScrollView API
  // Additional StandardTableView API
}
```

(a) is likely a bit more work, but probably the most pragmatic approach.

For (2) we could add interfaces that match the non-std-widgets base component interfaces, and the implementation would continue to be provided by the base component. That wouldn't change much at the moment - it would just define a contract for use in the future when we decide how to expose the interfaces for use in custom styles.

**Comment by 0x6e** [comment]
TabWidget seems like the style implementations could do with a bit of refactoring/consolidation, but I haven't delved deeply into all the implementations. Before I start that, were there any other planned changes there?

**Comment by ogoffart** [comment]
How would the implementation work when you also want the implementation to use inheritance?

Like MyStandardListView want to inherit from MyListView, but get the StandardListView interface from a StandardListViewBase element? would that work? is that desired?

**Comment by 0x6e** [comment]
You wouldn't be able to use the `uses` keyword because the ScrollView properties would conflict. Both base and MyListView are trying to provide the inheritedScrollView API:

```slint
component MyStandardListView uses { StandardListView from base } inherits MyListView {
//                                  >              <error{Cannot use interface 'StandardListView' because 'viewport-x' conflicts with an existing property in 'MyListView'}
    base := MyStandardListViewBase { }
}
```

You could use the `implements` keyword and manually redirect the StandardListView properties to base:

```slint
component MyStandardListView implements StandardListView inherits MyListView {
    current-item <=> base.current-item;
    // ...etc for the additional StandardTableView API

    base := MyStandardListViewBase { }
}
```

I think this is a bit of a contrived example - the problems with it don't go away if we use option (b): we would still get duplicated properties in the `uses` example.

I don't think either approach would affect how a user could provide an implementation of the interface. For me this is more a question of "is this the right time to deal with inheritance in interfaces?" That has been something we have deferred in the past, but now we need to provide interfaces for components that do actually inherit from one another, so it makes sense to revisit.

I think we should implement this via inheritance. One of the goals of the interfaces work is to reduce duplication of code. I don't think choosing to take the inheritance route has implications for how users would be able to implement the interface.

**Comment by 0x6e** [comment]
# Interface Review

We had a call on Mattermost to discuss interfaces as they are currently implemented, and the direction we would like them to take.

Attendees: @0x6e @eira-fransham @LeonMatthes @Murmele @NigelBreslaw @ogoffart @tronical

## Expected Usecases

The current implementation was primarily driven by a desire to:

- provide a single source of truth for the public std-widgets API;
- reduce boilerplate in the std-widgets style implementations;
- to provide a way to **check** that the std-widgets API has been meet by custom style implementations.

In the future we also intend to use interfaces to allow [named slots](https://github.com/slint-ui/slint/issues/6258#issuecomment-3879091189) to define the API that is expected by the component that declares the slot.

## Current Implementation

- An interface is declared with the `interface` keyword;
- An interface lists expected properties, callbacks, and function signatures;
- A component declares conformance via implements, separate from inheritance (inherits);
- Interface members may be inherited from a base component, declared explicitly, or auto-copied from the interface if missing;
- Interfaces can define default values/bindings, implementors may override them;
- Functions are the only members actually checked against the implementor — properties/callbacks just get defaults. Functions in interfaces must be implemented (no default function bodies; you can't override functions, but you can override callbacks);
- A uses statement forwards interfaces implemented by child components (composition-over-inheritance). The child may implicitly implement the interface.

### Soon-to-be-true

Interface inheritance (interface B inherits A) — implemented on a branch but not yet PR'd.

## Discussion

Leon observed that there are two styles of "interface" that we could use:

1. "explicit Duck typing" where the the `interface` declares the contract, the `component` must explicitly implement it, and the compiler **only checks** that that implementation matches the declaration;
2. "trait style" interfaces where the `interface` describes the contract and the `component` explicitly implements it **in addition to** the existing API exposed by the component.

The current implementation is a third style somewhere between these two that checks function declarations, adds missing properties and callbacks, and it is not clear whether the interface is added in addition to the component API.

### Adding vs Checking

The core issue resolved around whether an interface should "add" things, or only "check" them.

We definitely want checking. This means that we cannot provide default implementations of the interface, as it prevents us from catching errors:

```slint
interface ButtonInterface {
    in property <string> text;
}

// This is a contrived example. The `txt` property is a typo. The current implementation adds `text`
// for us. The error would not be apparent until the user wonders why the `text` property isn't
// working.
// If we require explicit implementation then the compiler would emit an error because `text` is not
// implemented.
component MyButton implementd ButtonInterface {
    in property <string> txt;

    Text {
        text: root.txt;
    }
}
```

There was some discussion about whether we want "trait style" interfaces, and how to handle disambiguations. The outcome was inconclusive and requirese further discussion. 

### `uses` Statements

It was generally agreed that the `uses` behaviour would work for both styles of interfaces. The syntax could be improved in the future. The current behaviour allows implicit duck typing: the forwarded component does not have to use the `implements` keyword. It was decided that this was undesirable, and should be fixed.

### Alternative Syntax

Eira proposed an alternate syntax:

```slint
export interface ButtonInterface {
    in property <string> text;
    in-out property <bool> checked;
    callback clicked;
    public function toggle();
}
```

```slint
// Explicit implementation syntax:
export component Button {

    implement ButtonInterface {
        text <=> root.text;
        checked <=> root.checked;
        clicked <=> root.clicked;

        public function toggle() {
            root.toggle();
        }
    }

    in property <string> text: "Button";
    in-out property <bool> checked: false;
    callback clicked;
    public function toggle() {
        self.checked = !self.checked;
    }

    Text {
        text: root.text;
    }
}
```

```slint
// Implicit implementation syntax:
export component Button {
    implement ButtonInterface <=> self;

    in property <string> text: "Button";
    in-out property <bool> checked: false;
    callback clicked;
    public function toggle() {
        self.checked = !self.checked;
    }

    Text {
        text: root.text;
    }
}
```

```slint
// `uses` equivalent
export component MyButton {
    implement ButtonInterface <=> button;

    Rectangle {
        background: lightsteelblue;

        button := Button { }
    }
}
```

This syntax requires further discussion.

## Other Concerns

Both Nigel and Eira raised concerns about the teachability of the interfaces in Slint. It should not be necessary users of Slint to understand "interface" concepts in order to produce good UI. The intefaces feature targets component developers who are producing library-style components to be consumed by others.

Eira shared their observations that Rust traits are easier to teach than C# interfaces/subclasses.

Martin proposed a built-in component interface (x/y/width/height) for generic slots.

## Next Steps

The current implementation was a result of implementing the path of least resistance for the st-widgets use case. We intend to make the following changes to discover how they affect the std-widgets usage:

- [x] Remove default property values in interfaces (#12326);
- [ ] ~`uses` keyword requires the forwarded component to explicitly implement the interface~;
- [x] Require explicit implementation of interfaces (#12597);

**Comment by 0x6e** [comment]
Summary from an impromptu Mattermost discussion.

*Plot twist*: duck-typing is back!

---

## Interface Declaration

The interface declaration has not changed:

```slint
export interface ButtonInterface {
    in property <string> text;
    in-out property <bool> checked;
    callback clicked;
    public function toggle();
}
```

### Rules

- no defaults;
- no function definitions.

## Implementing an Interface

- [x] the `implements` keyword will move out of the component declaration line into the element body as an `implement` statement;
  - this allows us to expand the `implement` syntax into a more explicit set of bindings in the future;
  - it also allows us to re-use the `implement` key word for `uses`;
  - it gives us more space for implementing multiple interfaces;
  - #12498
- [x] a component may implement multiple interfaces;
  - users could already effectively do this by chaining components inheritance with implements statements;
  - #12498

```slint
export component MyButton {
    implement ButtonInterface <=> self; // We explicitly declare that we implement an interface

    // The required properties must exist.
    // Missing or incorrect declarations of any of these would be an error.
    // API can be inherited from base components.
    in property <string> text: "Button";
    in-out property <bool> checked: false;
    callback clicked;
    public function toggle() {
        self.checked = !self.checked;
    }

    // Multiple interface implementations are permitted
    implement AccessibilityInterface <=> self;

    // AccessibilityInterface API...
}
```

### Forwarding an Interface

- [x] the `uses` keyword is replaced by the new `implement` syntax - the child element name is used instead of `self`;
  - #12498
- [x] the child element does not have to explicitly implement the interface, duck-typing is permitted;
  - requiring an additional `implement` statement in the element body duplicates the original statement in the root body;
  - #12498

```slint
export component FancyButton inherits Rectangle {

    implement ButtonInterface <=> inner; // requires `inner` to have everything ButtonInterface requires, replaces `uses`

    inner := Rectangle {
        implement ButtonInterface <=> self; // disallow for now - no problem with this really, but not useful atm. Would be useful with slots

        in property <string> text: "";
        in-out property <bool> checked;
        callback clicked;
        public function toggle() { }
    }

}
```

### Rules

- we cannot introduce the same property more than once
    - for now. In the future we may wish to allow interfaces to shadow properties/functions/callbacks from the main component but this also requires disambiguiation;
- we cannot implement the same interface more than once;
- we **can** implement more than one interface at a time; 
- do not allow `implement` in a child element for now;
    - for now. In the future we maye wish to allow this if it is useful for slots;

## Interface Inheritance

- [ ] interface inheritance is generally accepted as desirable;
  - a component must explicitly implement all the interfaces in the inheritance chain;
    - this is a defensive measure to prevent inheritance diamonds if we decide to allow a more Rust trait-like interface implementations in the future.
  - #13311
 - [ ] **New** - Should interfaces support `@deprecated` and `@shadowable` attributes?
   - The WIP widget interfaces do not have access to these, which means:
     - `@deprecated` API is not part of the new interface (e.g. no `viewport-x`, `-y`, `-width`, `-height` in `TextEdit` interface);
     - `@shadowable` is not carried by the generated API, so the Qt `ScrollView` implementation keeps `<=> self` instead of having aliases generated with `<=> internal`.

```slint
export interface ScrollView { }
export interface ListView inherits ScrollView { }
export interface StandardListView inherits ListView { }
export interface StandardTableView inherits ListView { }

component MyView { 

    implement StandardTableView {
        // Where does ListView/ScrollView come from?
    }
    implement StandardListView <=> self;


    // Therefore we must also 
    implement ScrollView <=> self;
    implement ListView <=> self;
}
```

**Comment by eira-fransham** [comment]
Maybe worth mentioning the reasons that we decided that `implement FooInterface <=> sub-component;` shouldn't require `sub-component` to implement `FooInterface`.

1. For consistency with `implement FooInterface <=> self;`. The `self` binding is introducing the implementation, and therefore we can't require `self` to already implement `FooInterface`. Requiring this for `sub-component` would be a strange inconsistency.
2. It's trivial to work around with `component MyFoo inherits Foo { implement FooInterface <=> self; }`, which means that resolving the issue of a sub-component not implementing `FooInterface` becomes needless busywork.
3. It makes `implement FooInterface <=> whatever;` always equivalent to `implement FooInterface { a <=> whatever.a; b <=> whatever.b; }`, no matter what `whatever` is. We don't want to implement that latter syntax yet, but it's something we likely want to implement eventually. The syntax where we list every member of the interface would not require `whatever` to implement the interface (since you're just referring to fields), so this gives yet another way in which resolving a compilation error due to `whatever` not implementing the interface becomes needless busywork.

Long story short, the drawbacks of requiring a delegated-to component to implement the interface outweigh the benefits.

**Comment by eira-fransham** [comment]
Reasoning for replacing `implements` before the braces of a component definition with `implement` statements:

1. It allows us to unify `implements` and `uses` to a single syntax
2. It extends better to future usecases (out of scope for the MVP), such as:
    - Implementing interfaces with overlapping member names
    - Implementing an interface on instantiation instead of definition
    - Implementing interfaces where fields are delegated to different sub-components, as opposed to the entire implementation being delegated to a single sub-component
3. We decided that for interface inheritance hierarchies, _all_ interfaces must be explicitly implemented, in order to resolve diamond inheritance. Therefore, putting all `implements` definitions at the top of the definition would start to get quite messy. This would also apply to a component which implements many interfaces, even without inheritance. One `implement` statement per interface means that they all get put one after another at the same indentation, with obvious formatting rules.

We decided _not_ to allow `implement FooInterface;` as sugar for `implement FooInterface <=> self;`, as with the potential future inclusion of explicit `implement FooInterface { .. }` syntax it would be unclear to a reader whether `implement FooInterface;` means `implement FooInterface <=> self;` or `implement FooInterface { }`. We may give `implement FooInterface;` a meaning in the future, but disallowing it now gives us more design space later.
